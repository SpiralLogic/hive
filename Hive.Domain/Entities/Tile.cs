using System.Collections.Generic;
using System.Linq;

namespace Hive.Domain.Entities
{
    public class Tile
    {
        public Tile(int id, int playerId, string content, IEnumerable<Coordinates> moves = null)
        {
            Id = id;
            PlayerId = playerId;
            Content = content;
            Moves = moves?.ToList() ?? new List<Coordinates>();
        }

        public int Id { get; }
        public int PlayerId { get; }
        public string Content { get; }
        public ICollection<Coordinates> Moves { get; }
    }
}
