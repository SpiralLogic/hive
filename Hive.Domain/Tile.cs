using System.Collections.Generic;

namespace Hive.Domain
{
    public class Tile
    {
        private int _count=0;

        public Tile(string owner, TextContent content, string color)
        {
            Id = (++_count).ToString();
            Owner = owner;
            Content = content;
            Color = color;
        }

        public string Id { get; }
        public string Owner { get; }
        public TextContent Content { get; }
        public string Color { get; }
        public ICollection<GameCoordinate> AvailableMoves { get; } = new List<GameCoordinate>();
    }
}