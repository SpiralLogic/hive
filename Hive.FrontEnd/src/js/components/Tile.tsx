import React from 'preact/compat';
import isEqual from 'react-fast-compare';

import { HexCoordinates, PlayerId, TileContent, TileId } from '../domain';
import { handleDrop } from '../handlers';
import {tileDragEmitter} from '../emitter/tile-drag-emitter';

const defaultProps = {
    tileDragEmitter: tileDragEmitter,
};

type Props = {
    id: TileId;
    content: TileContent;
    playerId: PlayerId;
    availableMoves: HexCoordinates[];
} & typeof defaultProps;

const getPlayerColor = (playerId: PlayerId) => {
    const playerColors = ['#85dcbc', '#f64c72'];
    return playerColors[playerId] || 'red';
};

function Tile(props: Props) {
    const { id, availableMoves, content, playerId, tileDragEmitter } = props;

    function handleDragStart() {
        tileDragEmitter.emit({ type: 'start', tileId: id, tileMoves: availableMoves });
    }

    function handleDragEnd() {
        tileDragEmitter.emit({ type: 'end', tileId: id, tileMoves: availableMoves });
    }

    const attributes = {
        title: content,
        style: { '--color': getPlayerColor(playerId) },
        className: 'hex tile',
        draggable: !!availableMoves.length,
        ondragstart: handleDragStart,
        ondragend: handleDragEnd,
        ondrop: handleDrop,
    };

    return (
        <div {...attributes}>
            <span>{content}</span>
        </div>
    );
}

Tile.displayName = 'Tile';
Tile.defaultProps = defaultProps;

const TileMemo = React.memo(Tile, isEqual);

export default TileMemo;
