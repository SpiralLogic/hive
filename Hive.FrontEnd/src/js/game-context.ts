import { createContext, useContext, useEffect, useMemo, useState } from 'preact/compat';
import { GameState, Hexagon, Move, MoveTile, Player } from './domain';
import { Engine } from './game-engine';

interface GameContext {
    players: Player[];
    hexagons: Hexagon[];
    moveTile: MoveTile;
}

export const useNewHiveContext = (): [boolean, GameContext?] => {
    const [loading, setLoading] = useState(true);
    const [gameState, setGameState] = useState<GameState>(undefined!);

    useEffect(() => {
        Engine.initialState()
            .then((state) => {
                setGameState(state);
                return setLoading(false);
            })
            .catch((reason) => {
                throw new Error(reason);
            });
    }, []);

    const moveTile = useMemo(() => (move: Move) => {
        Engine.moveTile(move)
            .then((state) => {
                return setGameState(state);
            })
            .catch((reason) => {
                throw new Error(reason);
            });
    }, [setGameState]);

    if (loading) return [loading];

    return [loading, { ...gameState, moveTile }];
};

export const HiveContext = createContext<GameContext>({
    hexagons: [],
    players: [],
    moveTile: () => undefined,
});

export const useHiveContext = () => useContext(HiveContext);
// @ts-ignore
HiveContext.displayName = 'Hive Context';
