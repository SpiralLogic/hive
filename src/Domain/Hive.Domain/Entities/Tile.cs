using System.Collections.Generic;

namespace Hive.Domain.Entities
{
    public sealed record Tile(int Id, int PlayerId, Creature Creature)
    {
        public ISet<Coords> Moves { get; init; } = new HashSet<Coords>();
    }
}