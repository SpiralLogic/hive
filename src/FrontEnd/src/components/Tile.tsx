import { FunctionComponent, h } from 'preact';
import { Tile } from '../domain';
import { TileAction } from '../utilities/hive-dispatcher';
import { addHiveEventListener, useClassReducer, useHiveDispatcher } from '../utilities/hooks';
import { deepEqual } from 'fast-equals';
import { handleDrop, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { memo } from 'preact/compat';
import { useEffect, useState } from 'preact/hooks';

const tileSelector = `[tabindex].tile`;
const cellSelector = `[tabindex].cell`;
const playerSelector = `[tabindex].name`;

const TileFC: FunctionComponent<Tile> = (tile: Tile) => {
  const { id, moves, creature, playerId } = tile;
  const [focus, setFocus] = useState(tileSelector);
  const [classes, setClassList] = useClassReducer(`player${playerId} hex tile`);
  const hiveDispatcher = useHiveDispatcher();

  const deselect = () => {
    if (!classes.includes('selected')) return;
    setClassList({ type: 'remove', class: 'selected' });
    hiveDispatcher.dispatch({ type: 'tileDeselected', tile: tile });
  };

  const select = () => {
    if (classes.includes('selected')) return;
    hiveDispatcher.dispatch({ type: 'tileClear' });
    setClassList({ type: 'add', class: 'selected' });
    hiveDispatcher.dispatch({ type: 'tileSelected', tile: tile });
  };

  addHiveEventListener<TileAction>('tileSelect', (event: TileAction) => {
    if (event.tile.id == id) select();
  });

  addHiveEventListener<TileAction>('tileDeselect', (event: TileAction) => {
    if (event.tile.id == id) deselect();
  });

  addHiveEventListener('tileClear', () => {
    deselect();
    setFocus('');
  });

  const handleDragStart = (event: DragEvent) => {
    event.stopPropagation();
    select();
    setClassList({ type: 'add', class: 'beforeDrag' });
    setTimeout(() => setClassList({ type: 'remove', class: 'beforeDrag' }), 1);
  };
  const handleDragEnd = () => {
    hiveDispatcher.dispatch({ type: 'tileClear' });
    hiveDispatcher.dispatch({ type: 'tileDropped', tile: tile });
  };
  const handleClick = (event: MouseEvent) => {
    event.stopPropagation();
    !classes.includes('selected') ? select() : deselect();
  };
  const handleKeyDown = (e: KeyboardEvent) => {
    if (handleKeyboardNav(e) || !isEnterOrSpace(e)) return;
    e.stopPropagation();
    if (!classes.includes('selected')) {
      select();
      setFocus(cellSelector);
    } else {
      deselect();
    }
  };
  const handleMouseLeave = (event: { currentTarget: HTMLElement }) => event.currentTarget.blur();

  useEffect(() => {
    if (!focus) return;
    const focusElement =
      document.querySelector<HTMLElement>(focus) || document.querySelector<HTMLElement>(playerSelector);
    focusElement?.focus();
    setFocus('');
  }, [focus]);

  const attributes = {
    title: creature,
    class: classes,
    draggable: !!moves.length,
    tabindex: moves.length ? 0 : undefined,
  };
  const handlers = moves.length
    ? {
        onclick: handleClick,
        ondragstart: handleDragStart,
        ondragend: handleDragEnd,
        onkeydown: handleKeyDown,
        onmouseleave: handleMouseLeave,
        ondrop: handleDrop,
      }
    : {
        ondrop: handleDrop,
      };
  return (
    <div {...attributes} {...handlers}>
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
        <use href="#hex" />
        <use class="creature" href={`#${creature.toLowerCase()}`} />
      </svg>
    </div>
  );
};
TileFC.displayName = 'Tile';
export default memo(TileFC, deepEqual);
