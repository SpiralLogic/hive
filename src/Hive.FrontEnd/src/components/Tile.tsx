import '../css/tile.css';

import { FunctionComponent } from 'preact';

import Hexagon from './Hexagon';
import { computed, Signal } from '@preact/signals';
import { useClassSignal } from '../hooks/useClassReducer';

type Properties = {
  selected?: boolean;
  creature?: string;
  classes: Signal<string>;
  classAction: ReturnType<typeof useClassSignal>[1];

  [rest: string]: unknown;
};

const Tile: FunctionComponent<Properties> = (properties) => {
  const { classes, children, classAction, selected, creature, ...rest } = properties;
  classAction.add('tile');
  if (creature) classAction.add(`${creature.toLowerCase()}`);
  if (selected) classAction.add('selected');

  const svgs = creature ? <use className={`creature`} href={`#${creature.toLowerCase()}`} /> : undefined;
  return (
    <Hexagon svgs={svgs} {...rest} classes={classes}>
      {children}
    </Hexagon>
  );
};
Tile.displayName = 'Tile';
export default Tile;
