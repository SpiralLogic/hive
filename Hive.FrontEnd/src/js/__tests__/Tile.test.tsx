import { RenderResult, render, cleanup, fireEvent } from '@testing-library/preact';
import React from 'preact/compat';
import Tile from '../components/Tile';
import TileDragEmitter, { TileDragEvent } from '../emitter/tile-drag-emitter';
import * as CTX from '../game-context';
import { Fragment } from 'preact';

const player1Tile = { id: 1, playerId: 1, content: 'ant', availableMoves: [{ q: 1, r: 1 }] };
const player2Tile = { id: 2, playerId: 0, content: 'fly', availableMoves: [] };
let tileDragEmitter: TileDragEmitter;
let moveTileSpy = jest.fn();

const tileJSX = (tileDragEmitter: TileDragEmitter) => (
    <Fragment>
        <Tile key="1" {...player1Tile} tileDragEmitter={tileDragEmitter}/>
        <Tile key="2" {...player2Tile} tileDragEmitter={tileDragEmitter}/>
    </Fragment>
);

let container: RenderResult;

beforeEach(() => {
    moveTileSpy = jest.fn().mockImplementation(() => Promise.resolve({ players: [], hexagons: [] }));
    jest.spyOn(CTX, 'useHiveContext').mockImplementation(() => ({ moveTile: moveTileSpy, hexagons: [], players: [] }));
    tileDragEmitter = new TileDragEmitter();
    container = render(tileJSX(tileDragEmitter));
});

describe('Tile Render', () => {
    test('tile color is the player\'s color', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        expect(tiles[0].style.getPropertyValue('--color')).toBe('#f64c72');
        expect(tiles[1].style.getPropertyValue('--color')).toBe('#85dcbc');
    });

    test('tile has content', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        expect(tiles[0].textContent).toBe('ant');
        expect(tiles[1].textContent).toBe('fly');
    });
});

describe('Tile drag and drop', () => {

    function simulateEvent (target: HTMLElement, type: string) {
        const preventDefault = jest.fn();
        const e = new MouseEvent(type, {bubbles:true});
        Object.assign(e, {preventDefault});
        fireEvent(target, e);

        return preventDefault;
    }

    const emitSpy = jest.fn();
    beforeEach(() => {
        emitSpy.mockClear();
        tileDragEmitter.add(emitSpy);
    });

    test('Tile is draggable when there are available moves', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        expect(tiles[0].attributes.getNamedItem('draggable')).toHaveProperty('value', 'true');
    });

    test('is *not* draggable when there are no moves available', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        expect(tiles[1].attributes.getNamedItem('draggable')).toHaveProperty('value', 'false');
    });

    test('on drag emits start event', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');

        fireEvent.dragStart(tiles[0]);
        const expectedEvent: TileDragEvent = { type: 'start', tileId: player1Tile.id, tileMoves: player1Tile.availableMoves };

        expect(emitSpy).toHaveBeenCalledWith(expectedEvent);
    });

    test('on dragEnd emits end event', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        fireEvent.dragEnd(tiles[0]);
        const expectedEvent: TileDragEvent = { type: 'end', tileId: player1Tile.id, tileMoves: player1Tile.availableMoves };

        expect(emitSpy).toHaveBeenCalledWith(expectedEvent);
    });

    test('default on drop is prevented', () => {
        const tile = document.querySelectorAll<HTMLDivElement>('.tile')[1];
        const preventDefault = simulateEvent(tile, 'drop');

        expect(preventDefault).toHaveBeenCalled();
    });
});
describe('Tile Snapshot', () => {
    test('matches current snapshot', () => {
        expect(render(tileJSX(tileDragEmitter)).baseElement).toMatchSnapshot();
    });
});

afterEach(() => {
    cleanup();
});
