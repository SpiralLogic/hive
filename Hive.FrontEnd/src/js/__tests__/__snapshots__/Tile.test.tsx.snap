// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile Snapshot matches current snapshot 1`] = `
<body>
  <div>
    <div
      class="hex tile"
      draggable="true"
      style="--color: #f64c72;"
      title="ant"
    >
      <span>
        ant
      </span>
    </div>
    <div
      class="hex tile"
      draggable="false"
      style="--color: #85dcbc;"
      title="fly"
    >
      <span>
        fly
      </span>
    </div>
  </div>
  <div>
    <div
      class="hex tile"
      draggable="true"
      style="--color: #f64c72;"
      title="ant"
    >
      <span>
        ant
      </span>
    </div>
    <div
      class="hex tile"
      draggable="false"
      style="--color: #85dcbc;"
      title="fly"
    >
      <span>
        fly
      </span>
    </div>
  </div>
</body>
`;
